lib/frontend/src/frontend.rs

```diff
@@ -571,10 +571,57 @@ impl<'a> FunctionBuilder<'a> {
         self.ins().call(libc_memcpy, &[dest, src, size]);
     }
 
+    /// Optimised memcpy for small copys.
+    pub fn emit_small_memcpy(&mut self,
+                             isa: &TargetIsa,
+                             dest: Value,
+                             src: Value,
+                             size: u64,
+                             dest_align: u8,
+                             src_align: u8)
+    {
+        const THRESHOLD: u64 = 4;
+
+        assert!(src_align % 2 == 0,
+                "Alignment value for `src` is not a power of two.");
+
+        assert!(dest_align % 2 == 0,
+                "Alignment value for `dest` is not a power of two.");
+
+        macro_rules! fallback {
+            () => {{
+                let size_value = self.ins().iconst(types::I64, size as i64);
+                self.call_memcpy(isa, dest, src, size_value);
+            }}
+        }
+
+        if src_align == dest_align {
+            let min = ::std::cmp::min(src_align, dest_align) as u64;
+            let access_size = greatest_divisible_power_of_two(min, size);
+            let load_and_store_amount = size / access_size;
+
+            if load_and_store_amount > THRESHOLD {
+                fallback!();
+                return
+            }
+
+            let flags = MemFlags::new();
+            flags.aligned();
+
+            for i in 0..load_and_store_amount {
+                let offset = (access_size * i) as i32;
+                let value = self.ins().load(types::I8, flags, src, offset);
```

Comment:
The type for the load here should be an integer type the size of `access_size`. You can use `Type::int` to get such a type.